cmd/clusterctl/pkg/client/cluster/move_test.go

```diff
@@ -0,0 +1,636 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/internal/test"
+)
+
+func TestObjectGraph_getDiscoveryTypeMetaList(t *testing.T) {
+	type fields struct {
+		proxy Proxy
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		want    []metav1.TypeMeta
+		wantErr bool
+	}{
+		{
+			name: "Return CRDs + ConfigMap & Secrets",
+			fields: fields{
+				proxy: test.NewFakeProxy().
+					WithObjs(
+						test.FakeCustomResourceDefinition("foo", "Bar", "v1", "v2"),
+						test.FakeCustomResourceDefinition("foo", "Baz", "v1"),
+					),
+			},
+			want: []metav1.TypeMeta{
+				{APIVersion: "foo/v1", Kind: "Bar"},
+				{APIVersion: "foo/v2", Kind: "Bar"},
+				{APIVersion: "foo/v1", Kind: "Baz"},
+				{APIVersion: "v1", Kind: "Secret"},
+				{APIVersion: "v1", Kind: "ConfigMap"},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gb := newObjectGraph(tt.fields.proxy)
+			got, err := gb.getDiscoveryTypes()
+			if (err != nil) != tt.wantErr {
+				t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if tt.wantErr {
+				return
+			}
+
+			// sort lists in a predictable way
+			sort.Slice(got, sortTypeMetaList(got))
+			sort.Slice(tt.want, sortTypeMetaList(tt.want))
+
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("got = %v, wantGraph %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func sortTypeMetaList(list []metav1.TypeMeta) func(i int, j int) bool {
+	return func(i, j int) bool {
+		return list[i].GroupVersionKind().String() < list[j].GroupVersionKind().String()
+	}
+}
+
+type wantGraphItem struct {
+	virtual    bool
+	dependants []string
+}
+
+type wantGraph struct {
+	nodes map[string]wantGraphItem
+}
+
+func assetGraph(t *testing.T, got *objectGraph, want wantGraph) {
+	if len(got.uidToNode) != len(want.nodes) {
+		t.Fatalf("got = %d nodes, wantGraph %d nodes", len(got.uidToNode), len(want.nodes))
+	}
+
+	for uid, wantNode := range want.nodes {
+		v, ok := got.uidToNode[types.UID(uid)]
+		if !ok {
+			t.Fatalf("failed to get node with uid = %s", uid)
+		}
+
+		if v.virtual != wantNode.virtual {
+			t.Errorf("node with uid = %s, got virtual = %t, wantGraph %t", uid, v.virtual, wantNode.virtual)
+		}
+
+		if len(v.dependents) != len(wantNode.dependants) {
+			t.Fatalf("node with uid = %s, got dependants = %d, wantGraph %d", uid, len(v.dependents), len(wantNode.dependants))
+		}
+
+		for _, wantDependants := range wantNode.dependants {
+			found := false
+			for k := range v.dependents {
+				if k.identity.UID == types.UID(wantDependants) {
+					found = true
+				}
+			}
+			if !found {
+				t.Errorf("node with uid = %s, failed to get dependants %s", uid, wantDependants)
+			}
+		}
+	}
+}
+
+func TestObjectGraph_addObj(t *testing.T) {
+	type args struct {
+		objs []*unstructured.Unstructured
+	}
+
+	tests := []struct {
+		name string
+		args args
+		want wantGraph
```

Comment:
It is possible but it requires to change `dependents map[*node]struct{}` into `dependents map[types.UID]*node` and the result is much more verbose.

If I think about what this will result when refactoring test cases with ~20 objects I'm not super happy, but If you think this improves readability I'm ok with it

e.g. with two objects
 
before:
```
want: wantGraph{
    nodes: map[string]wantGraphItem{
        "1": { // the object: not virtual (observed), without dependants
            virtual:    false,
            dependants: nil,
        },
        "2": { // the object owner: virtual (not yet observed), with 1 dependants
            virtual:    true,
            dependants: []string{"1"},
        },
    },
}
```

after 

```
want: &objectGraph{
    uidToNode: func() map[types.UID]*node {
        node1 := &node{ // the object: not virtual (observed), without dependents
            identity:   objectReference{APIVersion: "a/v1", Kind: "A", UID: "1", Namespace: "ns", Name: "foo"},
            dependents: make(map[types.UID]*node),
            virtual:    false,
        }
        node2 := &node{ // the object owner: virtual (not yet observed), with 1 dependents
            identity: objectReference{APIVersion: "b/v1", Kind: "B", UID: "2", Namespace: "ns", Name: "bar"},
            dependents: map[types.UID]*node{
                node1.identity.UID: node1,
            },
            virtual: true,
        }

        return map[types.UID]*node{
            "1": node1,
            "2": node2,
        }
    }(),
},
```

Reply:
I was thinking it would be simpler, but you've clearly shown your way is better. Let's keep it the way you have it. Thanks!